v2/controllers/egress_watcher.go

```diff
@@ -0,0 +1,198 @@
+	EgressPort int
+}
+
+// +kubebuilder:rbac:groups=coil.cybozu.com,resources=egresses,verbs=get;list;watch
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch
+
+// Reconcile implements Reconciler interface.
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *EgressWatcher) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	logger.Info("start reconciling egress")
+	eg := &coilv2.Egress{}
+	if err := r.Get(ctx, req.NamespacedName, eg); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "failed to get egress")
+		return ctrl.Result{}, err
+	}
+	if eg.DeletionTimestamp != nil {
+		return ctrl.Result{}, nil
+	}
+
+	pods := &corev1.PodList{}
+	err := r.Client.List(ctx, pods, client.MatchingFields{
+		constants.PodNodeNameKey: r.NodeName,
+	})
+	if err != nil {
+		logger.Error(err, "failed to list Pod")
+		return ctrl.Result{}, err
+	}
+
+	for _, pod := range pods.Items {
+		for k, v := range pod.Annotations {
+			if !strings.HasPrefix(k, constants.AnnEgressPrefix) {
+				continue
+			}
+
+			if k[len(constants.AnnEgressPrefix):] != eg.Namespace {
+				continue
+			}
+
+			// shortcut for the most typical case
+			if v == eg.Name {
+				// Do reconcile
+				if err := r.reconcileEgressClient(ctx, eg, &pod, &logger); err != nil {
+					logger.Error(err, "failed to reconcile Egress client pod")
+					return ctrl.Result{}, err
+				}
+				continue
+			}
+
+			for _, n := range strings.Split(v, ",") {
+				if n == eg.Name {
+					if err := r.reconcileEgressClient(ctx, eg, &pod, &logger); err != nil {
+						logger.Error(err, "failed to reconcile Egress client pod")
+						return ctrl.Result{}, err
+					}
+					continue
+				}
+			}
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *EgressWatcher) reconcileEgressClient(ctx context.Context, eg *coilv2.Egress, pod *corev1.Pod, logger *logr.Logger) error {
+	hook, err := r.getHook(ctx, eg, logger)
+	if err != nil {
+		return fmt.Errorf("failed to setup NAT hook: %w", err)
+	}
+
+	var ipv4, ipv6 net.IP
+	for _, podIP := range pod.Status.PodIPs {
+		ip := net.ParseIP(podIP.IP)
+		if ip.To4() != nil {
+			ipv4 = ip.To4()
+			continue
+		}
+		if ip.To16() != nil {
+			ipv6 = ip.To16()
+		}
+	}
+	if err := r.PodNet.Update(ipv4, ipv6, hook); err != nil {
+		return fmt.Errorf("failed to update NAT configuration: %w", err)
+	}
+
+	return nil
+}
+
+type gwNets struct {
+	gateway   net.IP
+	networks  []*net.IPNet
+	sportAuto bool
+}
+
+func (r *EgressWatcher) getHook(ctx context.Context, eg *coilv2.Egress, logger *logr.Logger) (nodenet.SetupHook, error) {
+	var gw gwNets
+	svc := &corev1.Service{}
+
+	if err := r.Get(ctx, client.ObjectKey{Namespace: eg.Namespace, Name: eg.Name}, svc); err != nil {
+		return nil, err
+	}
+
+	// as of k8s 1.19, dual stack Service is alpha and will be re-written
+	// in 1.20.  So, we cannot use dual stack services.
```

Comment:
Egress Watcher doesn't support dual stack services because of coild?
If only because of k8s, dual stacks are now on [stable](https://kubernetes.io/docs/concepts/services-networking/dual-stack/) and we can support I think.

Reply:
The dual-stack service is stable, but coil-egress doesn't support it.  This code is referenced from here(but slightly different).
https://github.com/cybozu-go/coil/blob/main/v2/runners/coild_server.go#L341-L342

I will rewrite the comment.